Documentation/design-docs/framework-version-resolution.md

```diff
@@ -0,0 +1,290 @@
+
+It is illegal to specify both `applyPatches` and `rollForward` in `.runtimeconfig.json` (counting both per-config and per-framework reference scopes together). It is OK to specify `applyPatches` in `.runtimeconfig.json` and `rollForward` through either CLI or env. variable.
+
+
+## Framework resolution
+The above described format and handling of settings on framework references will in the end produce a graph where the application is the first node and each dependent framework is a node. Each edge in the graph is a framework reference which has two attributes: the `version` and the `rollForward` setting.
+The goal of the framework resolution algorithm is to resolve any potentially conflicting framework references and to find actual available framework on disk which would match the framework references.
+
+### Framework reference conflict resolution
+If there are two references to the same framework name, then the host needs to resolve the potential conflict. The rules are:
+* Take the higher `version`
+* Validate that the reference with the lower `version` allows roll-forward to the higher version. If not, fail.
+* Take the more restrictive `rollForward` setting
+
+The check whether the roll-forward is allowed follows the rules described above in the list of available settings for `rollForward`.
+
+To select the more restrictive `rollForward` setting the values are ordered like this:
+1. `Disable`
+1. `LatestPatch`
+1. `Minor`
+1. `LatestMinor`
+1. `Major`
+1. `LatestMajor`
+
+Setting which is higher in this list (lower order number) is more restrictive than the one which is lower. `Disable` is the most restrictive (no roll-forward allowed). `LatestMajor` is the least restrictive (any other setting is compatible with it). `Minor` and `LatestMinor` are more restrictive then `Major` because they don't allow different major version.
+
+For example:
+In this example the two framework references are for the same framework name.
+
+| First framework reference | Second framework reference | Resolved framework reference |
+| ------------------------- | -------------------------- | ---------------------------- |
+| 2.1.0 Minor               | 2.2.0 Major                | 2.2.0 Minor                  |
+| 2.1.0 Minor               | 3.0.0 Minor                | failure                      |
+| 2.1.0 LatestMajor         | 3.0.0 Minor                | 3.0.0 Minor                  |
+| 2.1.0 LatestMajor         | 3.1.2 Disable              | 3.1.2 Disable                |
+
+To maintain backward compatibility, each framework reference will also have to carry `applyPatches` setting. Similar policy will be used to resolve conflicts. The more restrictive setting value will be used. So if one of the two framework references has `applyPatches=false` then the resolved framework reference will also have `applyPatches=false`.
+
+### Algorithm
```

Comment:
Thank you - this is the best part of this document! I would put this front and center and reference the area of the algorithm when talking about each trade off above.

Reply:
I would leave this section late in the document.  I think this may be a personality preference.  I prefer to discuss intent before implementation details.

Reply:
While I understand that for a hardcode developer, algorithm is all that matters, I would like to use this document with other audiences as well... so keeping this near the bottom :-)